fix: Improve auto-suggest for dataview task format

Yaozhuwa opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

# Description
I have optimized the function of auto-suggest in dataview task format, in dataview task format
1. removed the extra space when accepting the suggestion.
2. modified the effect of accepting the 'Enter' suggestion, now it will make the cursor jump out of the current task property.
3. fixed the duplicate words in the priority suggestion.

<!--- Describe your changes in detail -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
1. when I accept the day suggestion, it will add a extra space after date, see gif below
![iShot_2024-02-17_22 31 55](https://github.com/obsidian-tasks-group/obsidian-tasks/assets/28787183/b7d5fedf-25c4-4775-af91-64a6cc9ddd3b)

2. when I accept the 'Enter' suggestion, it will break my current task property and split the `]` to next line. see gif below
![iShot_2024-02-17_22 34 59](https://github.com/obsidian-tasks-group/obsidian-tasks/assets/28787183/e6ab01fd-99c1-4343-a8ac-ed0ee2d76a4c)

3. below is duplicate words bug in the priority suggestion when in dataview task format
<img width="793" alt="image" src="https://github.com/obsidian-tasks-group/obsidian-tasks/assets/28787183/3d846eb4-bbd4-4d6c-abd1-8b6a9d678a20">

4. My pull request solve the 3 issue above, which is shown below:
![iShot_2024-02-17_22 46 46](https://github.com/obsidian-tasks-group/obsidian-tasks/assets/28787183/e3f7010a-6c95-4e78-ac85-f5d9fd6d2e6a)

## How has this been tested?
For I only modify the auto-suggest action, I just test all type of suggestions, include priority, created date, due, schedule, repeat ...

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->

## Screenshots (if appropriate)

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

Changes visible to users:

- [x] **Bug fix** (prefix: `fix` - non-breaking change which fixes an issue)
- [ ] **New feature** (prefix: `feat` - non-breaking change which adds functionality)
- [ ] **Breaking change** (prefix: `feat!!` or `fix!!` - fix or feature that would cause existing functionality to not work as expected)
- [ ] **Documentation** (prefix: `docs` - improvements to any documentation content **for users**)
- [ ] **Sample vault** (prefix: `vault` - improvements to the [Tasks-Demo sample vault](https://github.com/obsidian-tasks-group/obsidian-tasks/tree/main/resources/sample_vaults/Tasks-Demo))
- [ ] **Contributing Guidelines** (prefix: `contrib` - any improvements to documentation content **for contributors** - see [Contributing to Tasks](https://publish.obsidian.md/tasks-contributing/))

Internal changes:

- [ ] **Refactor** (prefix: `refactor` - non-breaking change which only improves the design or structure of existing code, and making no changes to its external behaviour)
- [ ] **Tests** (prefix: `test` - additions and improvements to unit tests and the smoke tests)
- [ ] **Infrastructure** (prefix: `chore` - examples include GitHub Actions, issue templates)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My code follows the code style of this project and passes `yarn run lint`.
- [ ] My change requires a change to the documentation.
- [ ] I have [updated the documentation](https://publish.obsidian.md/tasks-contributing/Documentation/About+Documentation) accordingly.
- [x] My change has adequate [Unit Test coverage](https://publish.obsidian.md/tasks-contributing/Testing/About+Testing).

## Terms

<!--
By submitting this pull request, you must agree to follow our
[contributing guide](https://publish.obsidian.md/tasks-contributing) and
[Code of Conduct](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CODE_OF_CONDUCT.md).
Put an x in the boxes to confirm you agree.
-->

- [x] My contribution follow this project's [contributing guide](https://publish.obsidian.md/tasks-contributing)
- [x] I agree to follow this project's [Code of Conduct](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CODE_OF_CONDUCT.md)

claremacrae commented:
Hi, thank you for looking at this.

1. Having watched the videos, please could you have a look at this issue:

- https://github.com/obsidian-tasks-group/obsidian-tasks/issues/1014

I wonder whether you have actually fixed that issue for dataview format, and whether the fix should be applied to all formats?

2. The [docs](https://publish.obsidian.md/tasks/Editing/Auto-Suggest#Details) for the  ⏎ character say:

>  When the ⏎ item is shown at the top of the menu, it is given as a default option to enter a new line instead of choosing a suggestion. It is only shown when there is no concrete match.

From the videos you kindly supplied, it looks like it may not be doing that now?

claremacrae commented:
@kedestin, hello.

This is "for your information...." and also, if/when you have time, please could you review and test this? See also my comments above...

I wonder whether some of these changes are needed because I missed them whilst testing #2204?

Yaozhuwa commented:
> Hi, thank you for looking at this.
> 
> 1. Having watched the videos, please could you have a look at this issue:
> 
> * [Avoid trailing spaces in task creation #1014](https://github.com/obsidian-tasks-group/obsidian-tasks/issues/1014)
> 
> I wonder whether you have actually fixed that issue for dataview format, and whether the fix should be applied to all formats?
> 
> 2. The [docs](https://publish.obsidian.md/tasks/Editing/Auto-Suggest#Details) for the  ⏎ character say:
> 
> > When the ⏎ item is shown at the top of the menu, it is given as a default option to enter a new line instead of choosing a suggestion. It is only shown when there is no concrete match.
> 
> From the videos you kindly supplied, it looks like it may not be doing that now?

1. My PR is different from issue #1014. And It can't fix that issue. That issue calls for feature to remove extra space at the end of a line when press enter to next line.

2. It does change the ⏎ character's feature in dataview mode.  I realized that my previous PR would cause inconsistency in the behavior of dataview mode and emoji mode mode ⏎ symbols. So I updated my PR today. I removed the ⏎ symbol prompt in dataview mode, since dataview mode only prompts after typing `[`, and there is actually no need for the ⏎ symbol feature. In addition, I've optimized the auto-completion effect when auto-suggest is selected, which will now be more consistent with emoji mode. See the GIF below
![iShot_2024-02-18_20 51 38](https://github.com/obsidian-tasks-group/obsidian-tasks/assets/28787183/b2e1c798-2285-4095-8997-2e9150b27bb1)
I also tested it with Obsidian auto-completion of English symbols turned off, and it works perfectly as follows.
![iShot_2024-02-18_20 55 05](https://github.com/obsidian-tasks-group/obsidian-tasks/assets/28787183/fe40cc36-59b0-4d4e-9d4f-82e1ec4d350d)

Yaozhuwa commented:
@claremacrae

claremacrae commented:
> @claremacrae

Thank you for this.

I have not yet had time to test it throughly, so have not replied...

Thanks for adding/reinstating support for `()`. From that GIF, it looks like the ⏎ has been removed from the menu.

⏎ was provided to help users on phones without keyboards, and I would prefer it not to be removed.

Could ⏎ do something helpful like move the cursor to the end of the line and then add a new line?

If that is not possible, then the documentation should be updated to say that the ⏎ character is not included in the menu with dataview format.

I will make some comments on the code changes, from an initial review.

claremacrae commented on src/Suggestor/Suggestor.ts:
```diff
@@ -386,6 +427,50 @@ function isAnyBracketOpen(line: string, brackets: [opening_bracket: string, clos
+ * @returns The last open bracket in line among the given bracket pairs. If no such bracket exists, return null.
+ */
+function lastOpenBracket(line: string, brackets: [opening_bracket: string, closing_bracket: string][]): string | null {
+    if (brackets.length === 0) {
+        return null;
+    }
+    const numOpeningBrackets = Object.fromEntries(brackets.map(([open, _]) => [open, 0]));
+    const openingOf = Object.fromEntries(brackets.map(([open, close]) => [close, open]));
+    const openBracketsStack = [];
+    for (let i = 0; i < line.length; i++) {
+        const c = line[i];
+        if (c in numOpeningBrackets) {
+            numOpeningBrackets[c]++;
+            openBracketsStack.push({ bracket: c, idx: i });
+        } else if (c in openingOf) {
+            if (numOpeningBrackets[openingOf[c]] >= 1) {
+                for (let idx = openBracketsStack.length - 1; idx >= 0; idx--) {
+                    if (openBracketsStack[idx].bracket == openingOf[c]) {
+                        openBracketsStack.splice(idx, 1);
+                        break;
+                    }
+                }
+            }
+            numOpeningBrackets[openingOf[c]] = Math.max(0, numOpeningBrackets[openingOf[c]] - 1);
+        }
+    }
```
Please could that logic be protected with unit tests? It will be easy for a future maintainer to make a mistake, and manually testing against the examples in the comments is not practical.

Yaozhuwa replied:
I already add unit test and update the doc of Auto-Suggest.md.

claremacrae reviewed (changes requested):
Many thanks for this change.

Currently the only way to test it is to build the plugin and test manually - and that does not ensure that all the changes are tested. And it does not prevent the cod being accidentally broken in future.

So, please could unit tests be added for each of the changes in behaviour.

The best way to do this would be to:

- check out `main` and write tests that demonstrate each of the current behaviours that you do not like, and that need to be improved.
- attach that file here, as it will help me with testing and review
- and then update the tests so that they pass with the new behaviour

The existing tests are in:
https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/tests/Suggestor/Suggestor.test.ts

Thanks.

claremacrae commented:
The contributing guide has [information about the automated tests](https://publish.obsidian.md/tasks-contributing/Testing/About+Testing).

Yaozhuwa commented:
@claremacrae 
I already add unit test and update the doc of Auto-Suggest.md.

> ⏎ was provided to help users on phones without keyboards, and I would prefer it not to be removed.
> Could ⏎ do something helpful like move the cursor to the end of the line and then add a new line?
> If that is not possible, then the documentation should be updated to say that the ⏎ character is not included in the menu with dataview format.

For in dataview mode, suggestions will only appear within square brackets [] or parentheses () , so I think it's no need for ⏎  in dataview mode otherwise you'll get the wrong line breaks after opening brackets.

See in the [doc](https://publish.obsidian.md/tasks/Editing/Auto-Suggest#Intelligent+Auto-Suggest) :
> For example, suggestions will only appear within square brackets [] or parentheses () when using the [Dataview Task Format](https://publish.obsidian.md/tasks/Reference/Task+Formats/Dataview+Format#Bracketed%20inline%20fields).

claremacrae commented:
> The best way to do this would be to:
> 
> * check out `main` and write tests that demonstrate each of the current behaviours that you do not like, and that need to be improved.
> * attach that file here, as it will help me with testing and review
> * and then update the tests so that they pass with the new behaviour

Thanks for adding the tests.

I tried running them against the code before the PR, so I could see the changes in behaviour. I see they test a new implementation-detail function, so that won't work.

We will review the changes as soon as we can.

claremacrae commented:
There's one thing in the tests that I don't understand, which is why one of the current tests is still passing...

It's this one:

https://github.com/obsidian-tasks-group/obsidian-tasks/blob/9e89cef4dccbd837a23ac3efd2036472d42471cb/tests/Suggestor/Suggestor.test.ts#L146-L176

What it does is to generate a list of all suggestions, and then use [ApprovalTests](https://publish.obsidian.md/tasks-contributing/Testing/Approval+Tests) to save the results in these two files:

**A markdown table showing all suggestions for Emoji format:**
https://github.com/obsidian-tasks-group/obsidian-tasks/blob/9e89cef4dccbd837a23ac3efd2036472d42471cb/tests/Suggestor/Suggestor.test.auto-complete_with__emoji__symbols_show_all_suggested_text.approved.md

**And a markdown table showing all suggestions for dataview format:**
https://github.com/obsidian-tasks-group/obsidian-tasks/blob/9e89cef4dccbd837a23ac3efd2036472d42471cb/tests/Suggestor/Suggestor.test.auto-complete_with__dataview__symbols_show_all_suggested_text.approved.md

The question is, why is that test still passing in the PR for dataview format? It should be showing the the `return` suggestion has been removed, and the text displayed for Priorities has changed.

@Yaozhuwa Can you see why it hasn't changed, based on where your changes are, please?

claremacrae commented:
> @Yaozhuwa Can you see why it hasn't changed, based on where your changes are, please?

I've figured out the reason.

Before this PR, `Suggestor.ts` used these values from the `settings` parameter:

```bash
# Suggestor git:(a9441da0) grep 'settings\.' Suggestor.ts
        suggestions = suggestions.slice(0, settings.autoSuggestMaxItems);
        if (wordUnderCursor.length >= Math.max(1, _settings.autoSuggestMinMatch)) {
    if (matchingSuggestions.length === 0 && _settings.autoSuggestMinMatch === 0) return genericSuggestions;
        if (dateString.length < settings.autoSuggestMinMatch) {
        if (recurrenceString.length < settings.autoSuggestMinMatch) return [];
        const maxGenericDateSuggestions = settings.autoSuggestMaxItems / 2;
```

As of 4bcf0a172ccfee385c1233c7f52b6ac4828dc087 in this PR, the following values are used:

```bash
# Suggestor git:(a9441da0) grep 'settings\.' Suggestor.ts
            if (settings.taskFormat != 'dataview') {
        suggestions = suggestions.slice(0, settings.autoSuggestMaxItems);
    const dataviewMode = _settings.taskFormat === 'dataview';
                    _settings.taskFormat == 'dataview'
        if (wordUnderCursor.length >= Math.max(1, _settings.autoSuggestMinMatch)) {
    if (matchingSuggestions.length === 0 && _settings.autoSuggestMinMatch === 0) return genericSuggestions;
    const dataviewMode = settings.taskFormat === 'dataview';
        if (dateString.length < settings.autoSuggestMinMatch) {
    const dataviewMode = settings.taskFormat === 'dataview';
        if (recurrenceString.length < settings.autoSuggestMinMatch) return [];
        const maxGenericDateSuggestions = settings.autoSuggestMaxItems / 2;
```

Note that the number of references to `_settings.taskFormat === 'dataview';`.

However, the tests do not set `_settings.taskFormat`, as it was never needed before.

So currently, none of the tests dataview-specific new code.

**I will have a think about how best to address this.**

claremacrae reviewed (approved):
Super - thanks very much indeed for addressing these issues.

I see you made this change on your `main`, which may make it harder for you to merge later changes in to your fork.

A small tip to help you with any future pull requests on any projects (not just Tasks): it will be easier if you make a branch for your changes that will go in to pull requests.
